crates/repl_ui/src/repl_state.rs

```diff
@@ -82,39 +34,56 @@ impl Default for ReplState {
     }
 }
 
+#[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
+pub enum ReplAction<'a> {
+    Eval {
+        opt_mono: Option<MonomorphizedModule<'a>>,
+        problems: Problems,
+        opt_var_name: Option<String>,
+    },
+    Exit,
+    Help,
+    Nothing,
+}
+
 impl ReplState {
     pub fn new() -> Self {
         Self {
-            validator: InputValidator::new(),
             past_defs: Default::default(),
             past_def_idents: Default::default(),
             last_auto_ident: 0,
         }
     }
 
-    pub fn step(&mut self, line: &str, dimensions: Option<(usize, usize)>) -> Result<String, i32> {
-        let arena = Bump::new();
-
-        match parse_src(&arena, line) {
-            ParseOutcome::Empty => Ok(TIPS.to_string()),
+    pub fn step<'a>(
+        &mut self,
+        arena: &'a Bump,
+        line: &str,
+        target_info: TargetInfo,
+        palette: Palette,
+    ) -> ReplAction<'a> {
+        match parse_src(arena, line) {
```

Comment:
@rtfeldman I noticed that we seem to be parsing here, then descending into `next_action`, where we parse again! This doesn't seem necessary. Is that a mistake, or am I missing some subtlety?
If it's a subtlety then we should probably explain it in a comment.

Reply:
I refactored it to get rid of this, and it still seems to pass Wasm tests locally.
(I can't run the other REPL tests due to a problem I'm having with `-lSystem`)

Reply:
Yeah I think you're right, there was a redundant parse in there!